### ISSUE TITLE
EndAlignment and ConditionalAssigment cops disagree about shovel as assignment

### ISSUE THREAD

**Issue by seanwalbran** [rubocop-hq/rubocop#2743]
The Style/ConditionalAssignment cop [recognizes the shovel operator as an assigment](https://github.com/bbatsov/rubocop/blob/6ec674c09474af8d3a917fd23c0c16a7789bb3e6/lib/rubocop/cop/style/conditional_assignment.rb#L153) and flags nested shovel operations like the below for correction:

``` ruby
if foo.start_with? 'b'
  baz << true
else
  baz << false
end
# shovel_nested.rb:5:1: C: Use the return of the conditional for variable assignment and comparison.
```

However, the Lint/EndAlignment cop does not recognize the shovel as assignment, and so, when configured with the `AlignWith: variable` option, will flag this un-nested alignment as incorrect:

``` ruby
baz << if foo.start_with? 'b'
  true
else
  false
end
# shovel.rb:6:3: C: Style/IndentationWidth: Use 2 (not -5) spaces for indentation.
# shovel.rb:7:1: C: Style/ElseAlignment: Align else with if.
# shovel.rb:9:1: W: Lint/EndAlignment: end at 9, 0 is not aligned with if at 5, 7.
```

In contrast, it works if the assignment is an equals sign:

``` ruby
baz = if foo.start_with? 'b'
  true
else
  false
end
# no offenses
```

Full example sources and rubocop output: https://gist.github.com/seanwalbran/4700ae505c893cab345c

```
$ rubocop -V
0.36.0 (using Parser 2.3.0.2, running on ruby 2.2.4 x86_64-darwin14)
```

**Comment by slowjack2k** [comment]
The equals assignment does also not work the expected way.

When it is transformed from


```ruby

if 1 == a
  baz = 1
else
   baz =  3
end
```

to


```ruby
  baz = if 1 == a
              1
            else
              3
   end

```

Then I  Lint/EndAlignment with end at ..., ... is not aligned with if at ...

correct would be

```ruby
  baz = if 1 == a
            1
        else
            3
        end

```

**Comment by alexdowad** [comment]
@slowjack2k Thanks for pointing that out. We already have an open GH issue for that problem.

@seanwalbran What do you think about calling `<<` an "assignment"? Do you think that is desirable?

**Comment by seanwalbran** [comment]
@alexdowad At least from an EndAlignment perspective, I definitely do think it is desirable :smile: -- and apparently the ConditionalAssigment author(s) agree.  I understand that it could have different interpretations in different contexts, but fundamentally it modifies the left hand side based on the right hand side, and to me that counts as being assignmentlike.
 
Fwiw, we've been running with the below patch for a week now without seeing any undesirable side effects or surprises.
```ruby
module RuboCop
  class Node
    prepend(
      Module.new do
        def asgn_method_call?
          method_name == :<< || super
        end
      end
    )
  end
end
```

**Comment by chrisnicola** [comment]
This seems to be broken right now. `ConditionalAssignment` asks you to use `<<` but EndAlignment doesn't seem to work with it. Also EndAlignment has two confusing options to perhaps do the same thing `variable` and `start_of_line`. Neither works properly with `<<` despite the issue below saying it should.

https://github.com/bbatsov/rubocop/issues/2399

If you use `variable` it will complain about the indentation of the block and that end is not aligned. If you use `start_of_line` it will complain about the indentation of the block even for a `=` assignment.
